Add smartcore based ZKML example (#914)
risc0/risc0: examples/smartcore-ml/README.md

> # RISC Zero Rust SmartCore Machine Learning Template

Reviewer: We have a suggested README format, I'll dig up the link but in essence, in addition to any project specific sections, we suggest this outline:
* Intro
* Use Cases
* Project Organization
* More Resources
* And if you need any project specific sections, those too

This isn't intended to force all readmes into a cookie-cutter, but to the extent you can reorder/reorganize this readme to conform more to that structure, it will be cleaner to read (and maintain) for folks familiar with other examples.

Reviewer 2: Would it be OK to land and adjust this in a followup PR?

Reviewer: Yes

A list of the supported models with definitions for their respective types can be found below.  Note that {float} is a placeholder for either f32 or f64:

```
let model: DecisionTreeClassifier<{float}, u32, DenseMatrix<{float}>, Vec<u32>> = trained_model;
```

```
let model: RandomForestClassifier<{float}, u8, DenseMatrix<{float}>, Vec<u8>> = trained_model;
```

```
let model: GaussianNB<{float}, _, DenseMatrix<{float}>, _> = trained_model;
```

```
let model: LogisticRegression<{flaot}, u32, DenseMatrix<{float}>, Vec<u32>> = trained_model;
```

```
let model: LinearRegression< {float}, {float}, DenseMatrix<{float}>, Vec<{float}>> = trained_model;
```

```
let model: KNNClassifier< {float}, u32, DenseMatrix<{float}>, Vec<u32>, Euclidian<{float}>> = trained_model;
```

```
let model: KMeans<{float}, u8, DenseMatrix<{float}>, Vec<u8>> = trained_model;
```

```
let model: PCA<{float}, DenseMatrix<{float}>> = trained_model;
```

NOTE:  For Support Vector Machine classifiers and regression models, you MUST use the following smartcore fork and import into you cargo.toml file:

smartcore = { git = "https://github.com/Roee-87/smartcore.git", features = ["serde"]}

> Deserialization of SVC and SVR does not includes that SVCParametersmake field.  It must be added back into the model struct manually after deserialization.  The SVC and SVR fork have ammended the visbility of the model struct, making the parameters field public and thereby allowing the parameters field to be directly inserted into the model struct after dserialization.  You must insert the same parameters that were used when training the model.

Reviewer: Some proofreading suggestions. I don't know I corrected correctly to "The SVC and SVR forks have" ... perhaps it's "The SVC and SVR fork has"?
```suggestion
Deserialization of SVC and SVR does not includes that `SVCParametersmake` field.  It must be added back into the model struct manually after deserialization.  The SVC and SVR forks have amended the visbility of the model `struct`, making the parameters field public and thereby allowing the parameters field to be directly inserted into the model struct after deserialization.  You must insert the same parameters that were used when training the model.
```

```
let mut model: SVC<{float}, i32, DenseMatrix<{float}>, Vec<i32>> = deserialized_svc_trained;
let params = &SVCParameters::default().with_c(200.0).with_kernel(Kernels::linear());
model.parameters = Some(params);
```

```
let mut model: SVR< '_, {float}, DenseMatrix<{float}>, Vec<{float}>> = deserialized_model;
//make sure to use the same SVRParameters used to train the model
let params = &SVRParameters::default().with_eps(2.0).with_c(10.0).with_kernel(Kernels::linear());
model.parameters = params;
```

> SVM models support four kernels:  Gaussian (rbf), Linear, Polynomial, and Sigmoid.  Kernels are selected within the .with_kernel() function in the SVCParameters or SVRParameters.

Reviewer: ```suggestion
SVM models support four kernels:  Gaussian (rbf), Linear, Polynomial, and Sigmoid.  Kernels are selected within the `.with_kernel()` function in the `SVCParameters` or `SVRParameters`.
```

Kernel::linear()
Kernel::rbf()
Kernel::polynomial()

> Kernel::sigmoid()

Reviewer: I'm not sure what you're intending with these lines, but the status quo isn't it.

Author: Fixed

## Quick Start

First, make sure [rustup](https://rustup.rs) is installed. This project uses a [nightly](https://doc.rust-lang.org/book/appendix-07-nightly-rust.html) version of [Rust](https://doc.rust-lang.org/book/ch01-01-installation.html). The [`rust-toolchain`](rust-toolchain) file will be used by `cargo` to automatically install the correct version.

To build all methods and execute the method within the zkVM, run the following command:

```
cargo run
```
### Running proofs remotely on Bonsai

*Note: The Bonsai proving service is still in early Alpha; an API key is required for access. [Click here to request access].*

If you have access to the URL and API key to Bonsai you can run your proofs
remotely. To prove in Bonsai mode, invoke `cargo run` with two additional
environment variables:

```

> BONSAI_API_KEY="YOUR_API_KEY" BONSAI_API_URL="BONSAI_URL" cargo run

Reviewer: If you use `< >` then people who forget to substitute get shell errors rather than something potentially more mysterious.
```suggestion
BONSAI_API_KEY=<YOUR_API_KEY> BONSAI_API_URL=<BONSAI_URL> cargo run
```

Author: Removing this section.

```

[Click here to request access]: https://bonsai.xyz/apply

> ## How to create a project based on this template
> 
> 
> 
> 
> ## Contributor's Guide
> We welcome contributions to documentation and code via PRs and GitHub Issues on our [main repository](http://www.github.com/risc0) or any of our other repositories.
> 
> ## Video Tutorial
> 
> ## Questions, Feedback, and Collaborations
> We'd love to hear from you on [Discord](https://discord.gg/risczero) or [Twitter](https://twitter.com/risczero).
> 
> [Structure of a zkVM Application]: https://dev.risczero.com/zkvm/developer-guide/zkvm-app-structure

Reviewer: I think these are left over?
```suggestion
```

Author: Yes, I'll delete this section.
